src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala

```diff
@@ -135,34 +139,53 @@ private[impl] class InstanceTrackerActor(
         val op = InstanceOpProcessor.Operation(deadline, sender(), instanceId, instanceUpdateOp)
         updaterRef.forward(InstanceUpdateActor.ProcessInstanceOp(op))
 
-      case msg @ InstanceTrackerActor.StateChanged(ack) =>
-        val maybeChange: Option[InstanceChange] = ack.effect match {
-          case InstanceUpdateEffect.Update(instance, oldInstance, events) =>
-            updateApp(instance.runSpecId, instance.instanceId, newInstance = Some(instance))
-            Some(InstanceUpdated(instance, lastState = oldInstance.map(_.state), events))
+      case InstanceTrackerActor.StateChanged(ack) =>
+        import context.dispatcher
 
-          case InstanceUpdateEffect.Expunge(instance, events) =>
+        val maybeChange: Future[Option[InstanceChange]] = ack.effect match {
+          case effect @ InstanceUpdateEffect.Update(instance, _, _) =>
+            repository.store(instance) // first store to the repository
+              .map(_ => Some(effect))
+              .recoverWith(tryToRecoverRepositoryFailure(effect))
+              .map(trackerEffect => // update task tracked based on the repository update result
+                trackerEffect.flatMap(ef => {
+                  updateApp(ef.instance.runSpecId, ef.instance.instanceId, newInstance = Some(ef.instance))
```

Comment:
I don't think `updateApp` is threadsafe. Checkout the [implementation](https://github.com/mesosphere/marathon/blob/9daa627daec7d3b6cbb1cfcdade9699c5d30c951/src/main/scala/mesosphere/marathon/core/task/tracker/InstanceTracker.scala#L82).

Reply:
Yep, you're changing actor's state (`instancesBySpec`) from the future context which is a race condition. 
It'll make the logic more complex but you should pipe the `trackerEffect` to yourself and handle it in another receive match case. Don't forget to handle the failure case too (see [here](https://github.com/mesosphere/marathon/blob/master/src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala#L141) for example)
Mixing actors and futures only works nicely as long as future effects are stateless.